src/dune_pkg/opam_solver.ml

```diff
@@ -106,26 +115,123 @@ module Context_for_dune = struct
     ; candidates_cache
     ; available_cache
     ; constraints
+    ; latest_released_base_compiler_version
     }
   ;;
 
   let pp_rejection f = function
     | Unavailable -> Format.pp_print_string f "Availability condition not satisfied"
   ;;
 
+  let opam_version_compare a b = OpamPackage.Version.compare a b |> Ordering.of_int
+
+  (* Returns a bool which is true iff the following conditions are met:
+     - the version preference is [Newest]
+     - the given package is the ocaml compiler metapackage
+     - the version of the given package is later than the latest
+       version of the ocaml-base-compiler package lacking the
+       avoid-version flag *)
+  let opam_file_is_ocaml_newer_than_latest_base_compiler_if_prefer_newest
```

Comment:
I still find this code hard to read tbh. I think the conversion to booleans to define and order is what's making it hard to read. Will try to get rid of it.

Reply:
To clarify, @rgrinberg are you planning to do this work or are you asking me to take a look at it. Is there any harm in merging it as is and making some readability changes in a later PR? The build system team is planning to do some more developer interviews soon and this change is necessary for that work so it would help us if it could be merged soon.

Reply:
I will try to simplify the code a bit